beater/beater.go

```diff
@@ -43,17 +45,21 @@ func (bt *beater) Run(b *beat.Beat) error {
 		go bt.client.PublishAll(events)
 	}
 
-	server := newServer(bt.config, callback)
-	start(server, bt.config.SSL)
-	defer stop(server)
+	bt.server = newServer(bt.config, callback)
+	bt.errCh = start(bt.server, bt.config.SSL)
 
 	logp.Info("apm-server is running! Hit CTRL-C to stop it.")
-	// Blocks until service is shut down
-	<-bt.done
 
-	return nil
+	err = <-bt.errCh
+	logp.Err(err.Error())
+
+	if err == http.ErrServerClosed {
+		return nil
+	}
+	return err
 }
 
+// Graceful shutdown
 func (bt *beater) Stop() {
-	close(bt.done)
+	stop(bt.server)
```

Comment:
I would not recommend to use `stop` here but use `defer` instead. This has proven to be better on edge cases like when Stop is called before the server is actually started. Also `Stop` might potentially be removed in the future.